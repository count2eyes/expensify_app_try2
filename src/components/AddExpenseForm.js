import React from "react";
import { connect } from "react-redux";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";

export class AddExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      focused: false,
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? props.expense.amount : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      error: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;

    this.setState(() => ({ description }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (/^\d*(\.\d{0,2})?$/.test(amount)) {
      this.setState(() => ({ amount }));
    } else return undefined;
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onDateChange = date =>
    this.setState({
      createdAt: date
    });

  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide expense description & amount!"
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: this.state.amount,
        note: this.state.note,
        createdAt: moment(this.state.createdAt)
      });
    }
  };

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          {this.state.error ? <h4>{this.state.error}</h4> : null}
          <input
            type="text"
            onChange={this.onDescriptionChange}
            value={this.state.description}
            placeholder="Description"
          />
          <input
            type="number"
            onChange={this.onAmountChange}
            value={this.state.amount}
            placeholder="Amount"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={({ focused }) =>
              this.setState({
                focused
              })
            }
            isOutsideRange={() => false}
            numberOfMonths={1}
            id="your_unique_id"
          />
          <textarea
            onChange={this.onNoteChange}
            value={this.state.note}
            placeholder="Your expense note (optional)"
          />
          <button type="submit">Submit</button>
        </form>
      </div>
    );
  }
}
export default connect(state => state)(AddExpenseForm);
