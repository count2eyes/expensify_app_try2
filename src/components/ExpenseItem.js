import React from "react";
import moment from "moment";
import { Link } from "react-router-dom";
import numeral from "numeral";

const ExpenseItem = props => {
  const { description, amount, note, createdAt, id } = props.expense;
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>
        {numeral(amount / 100).format("$0,0.00")} -{" "}
        {moment(createdAt).format("ddd DD MMM, YYYY")} <small>{note}</small>
      </p>
    </div>
  );
};

export default ExpenseItem;
