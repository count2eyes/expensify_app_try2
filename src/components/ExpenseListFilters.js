import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import "react-dates/initialize";
import { setText, sortBy, setStartDate, setEndDate } from "../actions/filters";

export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      focusedInput: null
    };
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onTextChange = e => {
    const text = e.target.value;
    this.props.setText(text);
  };

  onSelect = e => {
    const selectedOption = e.target.value;
    if (selectedOption === "date") {
      this.props.sortBy("date");
    } else if (selectedOption === "amount") {
      this.props.sortBy("amount");
    }
  };

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select onChange={this.onSelect}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId="your_unique_start_date_id"
          endDate={this.props.filters.endDate}
          endDateId="your_unique_end_date_id"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={focusedInput => this.setState({ focusedInput })}
          isOutsideRange={() => false}
          numberOfMonths={1}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => {
    dispatch(setStartDate(startDate));
  },
  setEndDate: endDate => {
    dispatch(setEndDate(endDate));
  },
  setText: text => {
    dispatch(setText(text));
  },
  sortBy: selectedOption => {
    dispatch(sortBy(selectedOption));
  }
});

const mapStateToProps = state => ({ filters: state.filters });

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
