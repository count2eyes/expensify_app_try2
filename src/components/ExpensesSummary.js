import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import selector from "../selectors/expenses";
import expensesTotal from "../selectors/expensesTotal";

export class ExpensesSummary extends React.Component {
  render() {
    const string = this.props.count > 1 ? "expenses" : "expense";
    const totalDOM = numeral(this.props.total / 100).format("$0,0.00");

    return (
      <div>
        <h3>
          Viewing {this.props.count} {string} totalling {totalDOM}
        </h3>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const expenses = selector(state.expenses, state.filters);
  return {
    total: expensesTotal(expenses),
    count: expenses.length
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
