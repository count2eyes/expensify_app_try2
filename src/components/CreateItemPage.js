import React from "react";
import { connect } from "react-redux";
import AddExpenseForm from "./AddExpenseForm";
import { startAddExpense } from "../actions/expenses";

export class CreateItem extends React.Component {
  constructor(props) {
    super(props);
  }
  onSubmit = stateObj => {
    this.props.startAddExpense(stateObj);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h3>Add your expense here</h3>
        <AddExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(CreateItem);
