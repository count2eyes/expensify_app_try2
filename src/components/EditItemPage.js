import React from "react";
import AddExpenseForm from "./AddExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import { connect } from "react-redux";

export class EditItem extends React.Component {
  constructor(props) {
    super(props);
  }

  onSubmit = stateObj => {
    this.props.startEditExpense(this.props.expense.id, stateObj);
    this.props.history.push("/");
  };

  onClick = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h3>Edit your expense</h3>
        <AddExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    );
  }
}
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, stateObj) => dispatch(startEditExpense(id, stateObj)),

  startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditItem);
