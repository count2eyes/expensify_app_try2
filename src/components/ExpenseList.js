import React from "react";
import selector from "../selectors/expenses";
import { connect } from "react-redux";
import ExpenseItem from "./ExpenseItem";

export class ExpenseList extends React.Component {
  render() {
    return (
      <div>
        {this.props.expenses.length === 0 ? (
          <h3>No expense</h3>
        ) : (
          selector(this.props.expenses, this.props.filters).map(expense => (
            <ExpenseItem expense={expense} key={expense.id} />
          ))
        )}
      </div>
    );
  }
}

export default connect(state => state)(ExpenseList);
