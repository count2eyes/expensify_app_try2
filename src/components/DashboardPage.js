import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const Dashboard = () => (
  <div>
    <h3>Your expenses</h3>
    <ExpenseListFilters />
    <ExpensesSummary />
    <ExpenseList />
  </div>
);

export default Dashboard;
