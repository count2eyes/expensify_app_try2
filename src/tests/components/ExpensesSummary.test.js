import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";

test("should render correctly without no expense", () => {
  const wrapper = shallow(<ExpensesSummary total={0} count={0} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render correctly with expense data", () => {
  const wrapper = shallow(<ExpensesSummary count={7} total={200} />);
  expect(wrapper).toMatchSnapshot();
});
