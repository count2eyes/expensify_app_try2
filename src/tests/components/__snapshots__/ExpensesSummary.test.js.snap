// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with expense data 1`] = `
<div>
  <h3>
    Viewing 
    7
     
    expenses
     totalling 
    $2.00
  </h3>
</div>
`;

exports[`should render correctly without no expense 1`] = `
<div>
  <h3>
    Viewing 
    0
     
    expense
     totalling 
    $0.00
  </h3>
</div>
`;
